mautic/mautic
# Send email to user, owner or any email address as a campaign event

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Related user documentation PR URL | https://github.com/mautic/documentation/pull/207
| Related developer documentation PR URL | /
| Issues addressed (#s or URLs) | https://github.com/mautic/mautic/issues/437, partly https://github.com/mautic/mautic/issues/2433, https://github.com/mautic/mautic/issues/124
| BC breaks? | N (some translation strings moved around)
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
This PR will allow to send email to user, contact's owner or any To, Cc, Bcc email addresses as a campaign action.

The email tokens will get populated with the real values including contact field values. But the email hash is bogus so the links like unsubscribe won't work correctly. It's similar behaviour like when a user sends itself a test email.

#### Steps to test this PR:
1. Apply this PR
2. Clear the Mautic cache
3. Create a new campaign with your testing segment and "Send email" and "Send Email to User" actions. Test different configurations of the "Send Email to User" action.
4. Execute the `app/console mautic:campaigns:update` and `app/console mautic:campaigns:trigger` commands.
5. Check your inbox and see if you received the emails you'd expect based on the events config.
6. Make sure the form action "send results" still works since there were changes in it too.
7. Run the new tests: `phpunit --bootstrap vendor/autoload.php --configuration app/phpunit.xml --filter CampaignSubscriberTest`
